client/packages/common/src/ui/components/inputs/Autocomplete/AutocompleteWithPagination.tsx

```diff
@@ -0,0 +1,159 @@
+import React, { FC, PropsWithChildren, SyntheticEvent } from 'react';
+import {
+  Autocomplete as MuiAutocomplete,
+  AutocompleteRenderInputParams,
+  createFilterOptions,
+  styled,
+  Popper,
+  PopperProps,
+  CircularProgress,
+  Box,
+} from '@mui/material';
+import { BasicTextInput } from '../TextInput';
+import { useDebounceCallback } from '@common/hooks';
+import type { AutocompleteProps } from './Autocomplete';
+
+export interface AutocompleteWithPaginationProps<T>
+  extends AutocompleteProps<T> {
+  pagination?: {
+    page: number;
+    first: number;
+    offset: number;
+    total: number;
+  };
+  paginationDebounce?: number;
+  onPageChange?: (page: number) => void;
+}
+
+const StyledPopper = styled(Popper)(({ theme }) => ({
+  boxShadow: theme.shadows[2],
+}));
+
+export function AutocompleteWithPagination<T>({
+  defaultValue,
+  filterOptionConfig,
+  filterOptions,
+  getOptionDisabled,
+  optionKey,
+  loading,
+  loadingText,
+  noOptionsText,
+  onChange,
+  options,
+  renderInput,
+  renderOption,
+  width = 'auto',
+  value,
+  isOptionEqualToValue,
+  clearable = true,
+  disabled,
+  onInputChange,
+  inputValue,
+  autoFocus = false,
+  getOptionLabel,
+  popperMinWidth,
+  inputProps,
+  pagination,
+  paginationDebounce,
+  onPageChange,
+  ...restOfAutocompleteProps
+}: PropsWithChildren<AutocompleteWithPaginationProps<T>>) {
+  const filter = filterOptions ?? createFilterOptions(filterOptionConfig);
+
+  const defaultRenderInput = (props: AutocompleteRenderInputParams) => (
+    <BasicTextInput
+      {...props}
+      {...inputProps}
+      autoFocus={autoFocus}
+      InputProps={{
+        ...props.InputProps,
+        disableUnderline: false,
+        // style: props.disabled ? { paddingLeft: 0 } : {},
+        endAdornment: (
+          <>
+            {loading ? <CircularProgress color="primary" size={18} /> : null}
+            {props.InputProps.endAdornment}
+          </>
+        ),
+      }}
+      sx={{ width }}
+    />
+  );
+
+  const defaultRenderOption: FC = (
+    props: React.HTMLAttributes<HTMLLIElement>,
+    option: { id?: string; label?: string } & T
+  ) => (
+    <Box component="li" {...props} key={option.id}>
+      {option.label}
+    </Box>
+  );
+
+  const defaultGetOptionLabel = (option: T): string => {
+    if (!!optionKey) return String(option[optionKey]);
+
+    return (option as { label?: string }).label ?? '';
+  };
+
+  const debounceOnPageChange = useDebounceCallback(
+    (page: number) => {
+      onPageChange && onPageChange(page);
+    },
+    [onPageChange],
+    paginationDebounce || 500
```

Comment:
the `useDebounceCallback` function has a default of 500 anyway

```suggestion
    paginationDebounce
```